# 03.3 Data consistency checks on products, orders

### This script contains the following points:
### 1. Importing libraries and files
### 2. Data consistency checks on products file
#### 2.1 Check for mixed data types
#### 2.2 Check for missing values
#### 2.3 Check for duplicates
### 3. Data consistency checks on orders_wrangled file
#### 3.1 Check for mixed data
#### 3.2 Check for missing values
#### 3.3 Check for duplicates
### 4. Export data sets

# 1. Importing libraries and files

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Import csv files
path = r'C:\Users\dsadl\OneDrive\Documents\Career Foundry\Data Immersion\Project 4\05-10-2023 Instacart Basket Analysis'

In [3]:
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col=False)

In [4]:
df_prods = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'products.csv'), index_col=False)

# 2. Data consistency checks on products file

In [8]:
# Create a dataframe
df_test = pd.DataFrame()

In [9]:
# Create a mixed type column
df_test ['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

### 2.1 Check for mixed data types

In [11]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [12]:
# Convert from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

### 2.2 Check for missing values

In [5]:
# Check for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [6]:
# Display missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [7]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [8]:
# Identify number of rows in df_prods
df_prods.shape

(49693, 5)

In [9]:
# Create new df_prods dataframe without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [10]:
# Check number of rows in new dataframe to confirm creation of new dataframe
df_prods_clean.shape

(49677, 5)

### 2.3 Check for duplicates

In [11]:
# Check for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [12]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [13]:
# Create new dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [14]:
# Confirm by checking number of rows
df_prods_clean_no_dups.shape

(49672, 5)

In [15]:
# Export cleaned dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [31]:
df_prods_clean_no_dups.shape

(49672, 5)

# 3. Data consistency checks on orders_wrangled file

### 3.1 Check for mixed data types

In [16]:
# Initial check on file data
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week    order_time  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

#### The count for the "days_since_prior_order column" is less than the other columns, which may indicate missing values.

In [17]:
# The changes from the last exercise weren't reflected in the orders_wrangled.csv file. I deleted the "eval_count" columns and changed the data types of the "order_id", "user_id", and "order_number" columns
df_ords_3 = df_ords.drop(columns = ['eval_set','Unnamed: 0'])

In [18]:
df_ords_3['order_id'] = df_ords['order_id'].astype('str')

In [19]:
df_ords_3['user_id'] = df_ords['user_id'].astype('str')

In [20]:
df_ords_3['order_number'] = df_ords['order_number'].astype('str')

In [21]:
# Checking the success of the changes
df_ords_3.describe()

orders_day_of_week    order_time  days_since_prior_order
count        3.421083e+06  3.421083e+06            3.214874e+06
mean         2.776219e+00  1.345202e+01            1.111484e+01
std          2.046829e+00  4.226088e+00            9.206737e+00
min          0.000000e+00  0.000000e+00            0.000000e+00
25%          1.000000e+00  1.000000e+01            4.000000e+00
50%          3.000000e+00  1.300000e+01            7.000000e+00
75%          5.000000e+00  1.600000e+01            1.500000e+01
max          6.000000e+00  2.300000e+01            3.000000e+01

In [22]:
# Check for mixed data types
for col in df_ords_3.columns.tolist():
  weird = (df_ords_3[[col]].applymap(type) != df_ords_3[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords_3[weird]) > 0:
    print (col)

#### No mixed data types found


###  3.2 Check for missing values

In [23]:
# Check for missing values
df_ords_3.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_time                     0
days_since_prior_order    206209
dtype: int64

In [24]:
# Check if the missing values contain null values
df_nan_3 = df_ords_3[df_ords['days_since_prior_order'].isnull() == True]

In [25]:
df_nan_3.head(10)

order_id user_id order_number  orders_day_of_week  order_time  \
0   2539329       1            1                   2           8   
11  2168274       2            1                   2          11   
26  1374495       3            1                   1          14   
39  3343014       4            1                   6          11   
45  2717275       5            1                   3          12   
50  2086598       6            1                   5          18   
54  2565571       7            1                   3           9   
75   600894       8            1                   6           0   
79   280530       9            1                   1          17   
83  1224907      10            1                   2          14   

    days_since_prior_order  
0                      NaN  
11                     NaN  
26                     NaN  
39                     NaN  
45                     NaN  
50                     NaN  
54                     NaN  
75                     NaN  
79                     NaN  
83                     NaN

In [26]:
df_nan_3.shape

(206209, 6)

#### There are 206,209 missing values in the "days_since_prior_order" column. They contain 'NaN' values, as shown in creating a subset of the missing values that has 206,209 rows.

#### The best choice to resolve the null values in the "days_since_prior_order" column was to leave them as null values. All of these values had an order_number of "1", so the null values indicated it was the customers' first order. 

### 3.3 Check for duplicates

In [27]:
# Check for duplicates
df_dups = df_ords_3[df_ords_3.duplicated()]

In [28]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_time, days_since_prior_order]
Index: []

#### No full duplicates were found.

# 4. Export data sets

In [106]:
# Exporting dataframes
df_ords_3.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [29]:
df_ords_3.shape

(3421083, 6)

In [30]:
df_ords_3.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.pkl'))